geo/src/algorithm/interior_point.rs

```diff
@@ -0,0 +1,787 @@
+    euclidean_distance::EuclideanDistance, line_intersection::LineIntersection,
+    lines_iter::LinesIter, relate::Relate,
+};
+use crate::geometry::*;
+use crate::sweep::*;
+use crate::GeoFloat;
+
+/// Calculation of interior points.
+/// An interior point is a point that's guaranteed to intersect a given geometry, and will be
+/// strictly on the interior of the geometry if possible, or on the edge of the geometry has zero
+/// area. A best effort will additionally be made to locate the point reasonably centrally.
+///
+/// For polygons, this point is located by drawing a line that approximately subdivides the
+/// bounding box around the polygon in half, intersecting it with the polygon, then calculating
+/// the midpoint of the longest line produced by the intersection. For lines, the non-endpoint
+/// vertex closest to the line's centroid is returned if the line has interior points, or an
+/// endpoint is returned otherwise.
+///
+/// For multi-geometries or collections, the interior points of the constituent components are
+/// calculated, and one of those is returned (for MultiPolygons, it's the point that's the midpoint
+/// of the longest intersection of the intersection lines of any of the constituent polygons, as
+/// described above; for all others, the interior point closest to the collection's centroid is
+/// used).
+///
+/// # Examples
+///
+/// ```
+/// use geo::InteriorPoint;
+/// use geo::{point, polygon};
+///
+/// // rhombus shaped polygon
+/// let polygon = polygon![
+///     (x: -2., y: 1.),
+///     (x: 1., y: 3.),
+///     (x: 4., y: 1.),
+///     (x: 1., y: -1.),
+///     (x: -2., y: 1.),
+/// ];
+///
+/// assert_eq!(
+///     Some(point!(x: 1., y: 2.)),
+///     polygon.interior_point(),
+/// );
+/// ```
+pub trait InteriorPoint {
+    type Output;
+
+    /// See: <https://en.wikipedia.org/wiki/InteriorPoint>
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::InteriorPoint;
+    /// use geo::{line_string, point};
+    ///
+    /// let line_string = line_string![
+    ///     (x: 40.02f64, y: 116.34),
+    ///     (x: 40.02f64, y: 118.23),
+    ///     (x: 40.02f64, y: 120.15),
+    /// ];
+    ///
+    /// assert_eq!(
+    ///     Some(point!(x: 40.02, y: 118.23)),
+    ///     line_string.interior_point(),
+    /// );
+    /// ```
+    fn interior_point(&self) -> Self::Output;
+}
+
+impl<T> InteriorPoint for Line<T>
+where
+    T: GeoFloat,
+{
+    type Output = Point<T>;
+
+    fn interior_point(&self) -> Self::Output {
+        self.start_point()
+    }
+}
+
+impl<T> InteriorPoint for LineString<T>
+where
+    T: GeoFloat,
+{
+    type Output = Option<Point<T>>;
+
+    // The interior point of a LineString the non-endpoint vertex closest to the centroid if any,
+    // or the start point if there are no non-endpoint vertices
+    fn interior_point(&self) -> Self::Output {
+        match self.0.len() {
+            0 => None,
+            1 | 2 => Some(self.0[0].into()),
```

Comment:
This would have been my first thought too, but neither JTS nor Geos do this; they only ever return already-present vertices from lines and linestrings. I think this is because the baseline contract we're trying to upload here is:

```rust
if let Some(point) = geom.interior_point() {
    assert!(geom.intersects(&point)); // and also .contains(&point) if any such point exists
}
```

and for calculated midpoints, because of floating point rounding errors, this often isn't true in practice.

Consider this test script:
```rust
use geo::{Coordinate, Line, Intersects, Centroid};
use rand::Rng;

fn main() {
    let mut rng = rand::thread_rng();
    for _i in 0..100 {
        let start: Coordinate<f64> = (rng.gen(), rng.gen()).into();
        let end: Coordinate<f64> = (rng.gen(), rng.gen()).into();
        let line: Line<f64> = Line::new(start, end);
        let centroid = line.centroid();
        println!("{:?}", line.intersects(&centroid));
    }
}
```

in my testing, this only prints `true` about half the time.

All of which is to say: I think using actual vertices is indeed safer. FWIW, what JTS says they do is to use an interior point if any, or use whichever of the endpoints is closer to the centroid if not; this seemed sort of nonsensical to me because (modulo floating point weirdness again), they should be equidistant, so I'm just always using the start point. But if there's some situation where that doesn't do the right thing, I can calculate the distances I guess?

Reply:
Thank you for explaining! What you have makes sense. Could you add a comment to that effect?

Reply:
Now that I think about it, I suppose it's possible for this same issue to come up with degenerate flat/zero-area polygons that aren't horizontal or vertical, and it may be that perfectly correct behavior would need to detect that state of affairs and return vertices from the polygon instead of midpoints/intersection points. I don't think either Geos or JTS explicitly handle that case, though, so I'm probably not inclined to try either unless you feel strongly (I'm not sure it's something that could be cheaply detected).

Reply:
Added comments about lines

Reply:
Actually I lied; the polygon case isn't that hard to deal with either 😅 . Added a fallback to capture that case, and a new test.